### ISSUE TITLE
Authentication using Backdrop CMS

### ISSUE THREAD

**Issue by yorkshire-pudding** [meeting-room-booking-system/mrbs-code#32]
**Is your feature request related to a problem? Please describe.**
Backdrop CMS (https://backdropcms.org/) is a fork from Drupal 7 and the target market is small businesses and non-profits who need more comprehensive data driven sites.  The Drupal 7 MRBS module was removed many years ago as it was insecure.

I've looked at existing Auth modules, but I'm not quite sure what is required for Backdrop CMS (e.g. Joomla authentication requires no new modules on the Joomla site, just the roles).

**Describe the solution you'd like**
For users to be able to login to MRBS using their Backdrop CMS credentials
For roles for MRBS to be retrieved from Backdrop CMS
For users to be able to share the session of the Backdrop CMS site to avoid having to login to MRBS if they are already logged into 

**Describe alternatives you've considered**
Alternative CMS - they don't have the same power for joining data together
Separate logins - got this on existing D7 site.  Not ideal for a user journey.

**Additional context**

Documentation: https://docs.backdropcms.org/

**Comment by campbell-m** [comment]
I've had a very quick look at the Backdrop API.  It looks like the global $user object will give you the currently logged in user and that the user module should allow you to get details of other users.   What I don't know without more digging is how you load Backdrop from MRBS so that you get access to Backdrop globals and modules.

**Comment by yorkshire-pudding** [comment]
One possibility mentioned when I raised this in the backdrop community was the services module:
https://github.com/backdrop-contrib/services

**Comment by campbell-m** [comment]
That sounds as though it would be useful if MRBS was running on a remote machine. However I doubt it would give you details of the currently logged in user. It might allow you though to avoid MRBS and Backdrop users having to have two sets of credentials and for the admins to have to manage users in two separate places.

**Comment by yorkshire-pudding** [comment]
Doesn't sound as helpful as I thought then. Pretty much all the use cases I can think of would have MRBS on the same server and hosting account.

**Comment by yorkshire-pudding** [comment]
Hi @campbell-m - I'm looking at this again with the benefit of more experience in Backdrop. I'm also now the co-maintainer of a command line tool called `bee` which will probably be quite illuminating in terms of how it bootstraps backdrop so it can complete actions on it: see https://github.com/backdrop-contrib/bee

I'm interested in your steer on what needs to be done in MRBS to enable this.  I don't know how the other CMS integrations work (I've implemented MRBS with Joomla and it was on the same server so accessing the CMS was fine) but I don't think there would be any problem with limiting this to the Backdrop installation needing to be on same hosting account.

I'm happy to work on this with some guidance or simply provide input on backdrop functions if you prefer.

**Comment by yorkshire-pudding** [comment]
@campbell-m - I've done some experimenting but haven't managed to bootstrap Backdrop in a separate page.  Another approach could for MRBS to be installed as a module of Backdrop in a similar way to how CiviCRM is installed in Backdrop.
This would perhaps see it installed within:
`backdrop_root/modules/mrbs/backdrop/mrbs.info` and `backdrop_root/modules/mrbs/backdrop/mrbs.module`

That would ensure it has access to `$user` and could also get information for the users such as email address.

The Backdrop CMS website has CiviCRM so you can see how it is installed in this repo:
https://github.com/backdrop-ops/backdropcms.org/tree/main/www/modules/civicrm

Do you think this approach could work with MRBS?

**Comment by yorkshire-pudding** [comment]
Hi @campbell-m - I've figured this out and been able to load the current user in a test page within the MRBS folder.
I would like to have a go at building this integration which I think should mostly be on the MRBS side.  Please can you confirm my understanding of where I may need to add code (from the mrbs document root):

- `AUTHENTICATION`
- `INSTALL` (if needed to mention around multisites)
- possibly `web/grab_globals.inc.php` (based on presence of Wordpress code)
- possibly `web/lib/MRBS/Form/Form.php` (based on presence of Joomla code)
- `web/init.inc`
- `web/systemdefaults.inc.php`
- `web/auth/cms/backdropcms.inc` - NEW
- `web/lib/MRBS/Auth/AuthBackdropcms.php` - NEW
- `web/lib/MRBS/Session/SessionBackdropcms.php` - NEW
- if any other class extensions needed to go in `web/lib/MRBS` - should the names of these follow the class names

I'm happy if you want to assign me to this.  Is it best to ask for support here (if needed), or is there somewhere else for developers to collaborate?

**Comment by campbell-m** [comment]
Yes, please, that would be great if you could add support for Backdrop.  Could you add it as a Pull Request to the main branch please?

Yes, the summary above looks correct, though I think you may be working off MRBS 1.11.0 rather than the latest development code in the main branch.  For example, I've moved the Joomla specific code out of Form.php where it didn't really belong.

> if any other class extensions needed to go in web/lib/MRBS - should the names of these follow the class names

Yes, and if there are lots put them in a sub-directory.

I wonder whether it would be simpler just to call it 'Backdrop' rather than 'Backdropcms' to be consistent with Joomla and Wordpress?  I don't know how people normally refer to it.

> Is it best to ask for support here (if needed), or is there somewhere else for developers to collaborate?

You can ask for support here, email me directly or else there's an MRBS Slack workspace.  May need @jberanek to add you.

**Comment by yorkshire-pudding** [comment]
I think it might make sense to use Backdrop and backdrop within the code.  That would mirror how we code things in Backdrop but can we refer to it as Backdrop CMS in any documentation? This helps to ensure people can find it as searching for backdrop on the wider web is too vague.

Yes, I tried the latest release when troubleshooting #80 but was normally using the latest dev branch.  Yes, no problem to do a PR to the main branch when I've got something to test.

Out of interest, and as I would like to think about future proofing where possible, what should I consider with regards to the roles branch? How will the roles change?

**Comment by campbell-m** [comment]
> I think it might make sense to use Backdrop and backdrop within the code. That would mirror how we code things in Backdrop but can we refer to it as Backdrop CMS in any documentation?

OK.

> Out of interest, and as I would like to think about future proofing where possible, what should I consider with regards to the roles branch? How will the roles change?

I wouldn't worry about that for the moment.  There's still some thinking to be done there.
